File: web/src/app/features/repair-shop/data/schemas/repair-shop.schema.ts
import { z } from 'zod';

import { AuthUserSchema } from '../../../auth/domain/entities/auth-user';

export const RepairShopFormSchema = z.object({
  name: z
    .string()
    .trim()
    .min(5, 'El nombre del taller debe tener al menos 5 caracteres')
    .max(70, 'El nombre del taller debe tener maximo 70 caracteres'),
  latitude: z.number().min(-90, 'Latitud invalida').max(90, 'Latitud invalida'),
  longitude: z.number().min(-180, 'Longitud invalida').max(180, 'Longitud invalida'),
});

export const RepairShopSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  text_address: z.string(),
  latitude: z.number(),
  longitude: z.number(),
});

export const ServiceSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
});

export const ServicesListResponseSchema = z.array(ServiceSchema);

export const AssignShopServicesFormSchema = z.object({
  service_ids: z
    .array(z.string().uuid('Servicio invalido'))
    .min(1, 'Selecciona al menos un servicio'),
});

export const UpdateShopProfileFormSchema = z.object({
  name: z
    .string()
    .trim()
    .min(5, 'El nombre del taller debe tener al menos 5 caracteres')
    .max(70, 'El nombre del taller debe tener maximo 70 caracteres'),
  text_address: z
    .string()
    .trim()
    .min(5, 'La direccion debe tener al menos 5 caracteres')
    .max(240, 'La direccion debe tener maximo 240 caracteres'),
});

export const UpdateShopLocationFormSchema = z.object({
  latitude: z.number().min(-90, 'Latitud invalida').max(90, 'Latitud invalida'),
  longitude: z.number().min(-180, 'Longitud invalida').max(180, 'Longitud invalida'),
});

export const ShopInvitationCreateFormSchema = z.object({
  expires_at: z.string().min(1, 'Fecha de expiracion invalida'),
});

export const ShopInvitationSchema = z.object({
  code: z.string(),
  expires_at: z.string(),
  expires_at_label: z.string(),
  status: z.enum(['active', 'expired']),
});

export const RepairShopResponseSchema = z.object({
  user: AuthUserSchema,
  shop: RepairShopSchema,
});

export const ShopMechanicSchema = z.object({
  id: z.string().uuid(),
  shop_id: z.string().uuid(),
  user_id: z.string().uuid(),
  is_available: z.boolean(),
  created_date: z.string(),
  user: AuthUserSchema,
});

export const ShopMechanicsListResponseSchema = z.array(ShopMechanicSchema);

export const AdminRepairShopSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  text_address: z.string(),
  latitude: z.number(),
  longitude: z.number(),
  is_available: z.boolean(),
  state: z.boolean(),
  owner_id: z.string().uuid(),
  owner_name: z.string(),
  owner_email: z.string().email(),
  created_date: z.string(),
  deleted_date: z.string().nullable().optional(),
});

export const AdminRepairShopsResponseSchema = z.object({
  shops: z.array(AdminRepairShopSchema),
});

export const AdminShopMechanicStatsSchema = z.object({
  total: z.number(),
  available: z.number(),
  unavailable: z.number(),
  active_records: z.number(),
  inactive_records: z.number(),
});

export const AdminRepairShopOverviewSchema = z.object({
  shop: AdminRepairShopSchema,
  mechanic_stats: AdminShopMechanicStatsSchema,
  recent_activity: z.array(z.string()),
});

export const AdminRecentServiceSchema = z.object({
  assignment_id: z.string().uuid(),
  incident_id: z.string().uuid(),
  incident_description: z.string().nullable().optional(),
  incident_address: z.string().nullable().optional(),
  incident_status: z.string(),
  problem_name: z.string().nullable().optional(),
  mechanic_name: z.string().nullable().optional(),
  accepted_at: z.string().nullable().optional(),
  created_date: z.string(),
});

export const AdminRecentServicesResponseSchema = z.object({
  services: z.array(AdminRecentServiceSchema),
});

export type RepairShopFormData = z.infer<typeof RepairShopFormSchema>;
export type RepairShopData = z.infer<typeof RepairShopSchema>;
export type ServiceData = z.infer<typeof ServiceSchema>;
export type ShopInvitationData = z.infer<typeof ShopInvitationSchema>;
export type RepairShopResponseData = z.infer<typeof RepairShopResponseSchema>;
export type ShopMechanicData = z.infer<typeof ShopMechanicSchema>;
export type AdminRepairShopData = z.infer<typeof AdminRepairShopSchema>;
export type AdminShopMechanicStatsData = z.infer<typeof AdminShopMechanicStatsSchema>;
export type AdminRepairShopOverviewData = z.infer<typeof AdminRepairShopOverviewSchema>;
export type AdminRecentServiceData = z.infer<typeof AdminRecentServiceSchema>;

File: server/app/modules/repair_shop/schemas/repair_shop_schema.py
from datetime import datetime, timezone
from typing import Annotated, Literal
from uuid import UUID

from pydantic import StringConstraints, field_validator
from sqlmodel import SQLModel

from app.modules.user.schemas import UserRead


class RepairShopCreate(SQLModel):
    name: Annotated[str, StringConstraints(min_length=5, max_length=70)]
    latitude: float
    longitude: float

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if len(normalized) < 5:
            raise ValueError("El nombre del taller debe tener al menos 5 caracteres")
        return normalized

    @field_validator("latitude")
    @classmethod
    def validate_latitude(cls, value: float) -> float:
        if value < -90 or value > 90:
            raise ValueError("La latitud debe estar entre -90 y 90")
        return value

    @field_validator("longitude")
    @classmethod
    def validate_longitude(cls, value: float) -> float:
        if value < -180 or value > 180:
            raise ValueError("La longitud debe estar entre -180 y 180")
        return value


class RepairShopRead(SQLModel):
    id: UUID
    name: str
    text_address: str
    latitude: float
    longitude: float
    model_config = {"from_attributes": True}


class RepairShopAddressPreviewRead(SQLModel):
    name: str
    latitude: float
    longitude: float
    text_address: str


class RepairShopCreateResponse(SQLModel):
    user: UserRead
    shop: RepairShopRead


RepairShopOnboardingResponse = RepairShopCreateResponse


class ServiceRead(SQLModel):
    id: UUID
    name: str

    model_config = {"from_attributes": True}


class ShopMechanicRead(SQLModel):
    id: UUID
    shop_id: UUID
    user_id: UUID
    is_available: bool
    created_date: datetime
    user: UserRead


class AdminRepairShopRead(SQLModel):
    id: UUID
    name: str
    text_address: str
    latitude: float
    longitude: float
    is_available: bool
    state: bool
    owner_id: UUID
    owner_name: str
    owner_email: str
    created_date: datetime
    deleted_date: datetime | None = None


class AdminRepairShopsResponse(SQLModel):
    shops: list[AdminRepairShopRead]


class AdminShopMechanicStatsRead(SQLModel):
    total: int
    available: int
    unavailable: int
    active_records: int
    inactive_records: int


class AdminRepairShopOverviewRead(SQLModel):
    shop: AdminRepairShopRead
    mechanic_stats: AdminShopMechanicStatsRead
    recent_activity: list[str]


class AdminRecentServiceRead(SQLModel):
    assignment_id: UUID
    incident_id: UUID
    incident_description: str | None = None
    incident_address: str | None = None
    incident_status: str
    problem_name: str | None = None
    mechanic_name: str | None = None
    accepted_at: datetime | None = None
    created_date: datetime


class AdminRecentServicesResponse(SQLModel):
    services: list[AdminRecentServiceRead]


class RepairShopDashboardKpisRead(SQLModel):
    requests_received: int
    accepted_services: int
    completed_services: int
    cancelled_cases: int
    acceptance_rate: float
    cancellation_rate: float
    revenue_total: float
    average_resolution_minutes: float
    average_ticket_value: float
    top_mechanic_name: str | None = None
    top_mechanic_completed: int = 0


class RepairShopDashboardBreakdownRead(SQLModel):
    label: str
    count: int
    percentage: float


class RepairShopDashboardZoneRead(SQLModel):
    label: str
    count: int
    percentage: float
    latitude: float | None = None
    longitude: float | None = None


class RepairShopDashboardRead(SQLModel):
    period_days: int
    generated_at: datetime
    kpis: RepairShopDashboardKpisRead
    services_by_type: list[RepairShopDashboardBreakdownRead]
    zones_by_services: list[RepairShopDashboardZoneRead]


class ShopServicesAssignRequest(SQLModel):
    service_ids: list[UUID]

    @field_validator("service_ids")
    @classmethod
    def validate_service_ids(cls, value: list[UUID]) -> list[UUID]:
        unique_ids = list(dict.fromkeys(value))
        if len(unique_ids) == 0:
            raise ValueError("Debes seleccionar al menos un servicio")
        return unique_ids


class RepairShopProfileUpdateRequest(SQLModel):
    name: Annotated[str, StringConstraints(min_length=5, max_length=70)]
    text_address: Annotated[str, StringConstraints(min_length=5, max_length=240)]

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if len(normalized) < 5:
            raise ValueError("El nombre del taller debe tener al menos 5 caracteres")
        return normalized

    @field_validator("text_address")
    @classmethod
    def normalize_text_address(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if len(normalized) < 5:
            raise ValueError("La direccion debe tener al menos 5 caracteres")
        return normalized


class RepairShopLocationUpdateRequest(SQLModel):
    latitude: float
    longitude: float

    @field_validator("latitude")
    @classmethod
    def validate_latitude(cls, value: float) -> float:
        if value < -90 or value > 90:
            raise ValueError("La latitud debe estar entre -90 y 90")
        return value

    @field_validator("longitude")
    @classmethod
    def validate_longitude(cls, value: float) -> float:
        if value < -180 or value > 180:
            raise ValueError("La longitud debe estar entre -180 y 180")
        return value


class ShopInvitationCreateRequest(SQLModel):
    expires_at: datetime

    @field_validator("expires_at")
    @classmethod
    def validate_expires_at(cls, value: datetime) -> datetime:
        now_utc = datetime.now(timezone.utc)

        if value.tzinfo is None:
            value = value.replace(tzinfo=timezone.utc)

        if value <= now_utc:
            raise ValueError("La fecha de expiracion debe ser futura")

        return value


class ShopInvitationRead(SQLModel):
    code: str
    expires_at: datetime
    expires_at_label: str
    status: Literal["active", "expired"]


class JoinShopByCodeRequest(SQLModel):
    code: Annotated[str, StringConstraints(min_length=6, max_length=6)]

    @field_validator("code")
    @classmethod
    def validate_code(cls, value: str) -> str:
        normalized = value.strip().upper()
        if len(normalized) != 6:
            raise ValueError("El codigo debe tener 6 caracteres")
        if not normalized.isalnum():
            raise ValueError("El codigo solo permite letras y numeros")
        return normalized
